OpenRA.Mods.Common/Widgets/TextFieldWidget.cs

```diff
@@ -115,82 +120,180 @@ public override bool HandleKeyPress(KeyInput e)
-				return true;
+				case Keycode.ESCAPE:
+					if (OnEscKey())
+						return true;
+					break;
 
-			if (e.Key == Keycode.LALT && OnAltKey())
-				return true;
+				case Keycode.LALT:
+					if (OnAltKey())
+						return true;
+					break;
 
-			if (e.Key == Keycode.LEFT)
-			{
-				if (CursorPosition > 0)
-					CursorPosition--;
+				case Keycode.LEFT:
+					ResetBlinkCycle();
+					if (CursorPosition > 0)
+						if ((!isOSX && e.Modifiers.HasModifier(Modifiers.Ctrl)) ||
+						(isOSX && e.Modifiers.HasModifier(Modifiers.Alt)))
+							CursorPosition = getPrevWhitespaceIndex();
+						else
+							CursorPosition--;
+					break;
 
-				return true;
-			}
+				case Keycode.RIGHT:
+					ResetBlinkCycle();
+					if (CursorPosition <= Text.Length - 1)
+						if ((!isOSX && e.Modifiers.HasModifier(Modifiers.Ctrl)) ||
+						(isOSX && e.Modifiers.HasModifier(Modifiers.Alt)))
+							CursorPosition = getNextWhitespaceIndex();
+						else
+							CursorPosition++;
 
-			if (e.Key == Keycode.RIGHT)
-			{
-				if (CursorPosition <= Text.Length - 1)
-					CursorPosition++;
+					break;
 
-				return true;
-			}
+				case Keycode.HOME:
+					ResetBlinkCycle();
+					CursorPosition = 0;
+					break;
 
-			if (e.Key == Keycode.HOME)
-			{
-				CursorPosition = 0;
-				return true;
-			}
+				case Keycode.END:
+					ResetBlinkCycle();
+					CursorPosition = Text.Length;
+					break;
 
-			if (e.Key == Keycode.END)
-			{
-				CursorPosition = Text.Length;
-				return true;
-			}
+				case Keycode.K:
+					// ctrl+k is equivalent to cmd+delete on osx
+					ResetBlinkCycle();
+					if (!isOSX && e.Modifiers.HasModifier(Modifiers.Ctrl) && CursorPosition < Text.Length)
+					{
+						Text = Text.Remove(CursorPosition);
+						OnTextEdited();
+					}
 
-			if (e.Key == Keycode.DELETE)
-			{
-				if (CursorPosition < Text.Length)
-				{
-					Text = Text.Remove(CursorPosition, 1);
-					OnTextEdited();
-				}
+					break;
 
-				return true;
-			}
+				case Keycode.U:
+					// ctrl+u is equivalent to cmd+backspace on osx
+					ResetBlinkCycle();
+					if (!isOSX && e.Modifiers.HasModifier(Modifiers.Ctrl) && CursorPosition > 0)
+					{
+						Text = Text.Substring(CursorPosition);
+						CursorPosition = 0;
+						OnTextEdited();
+					}
 
-			if (e.Key == Keycode.BACKSPACE && CursorPosition > 0)
-			{
-				CursorPosition--;
-				Text = Text.Remove(CursorPosition, 1);
-				OnTextEdited();
-				return true;
-			}
+					break;
 
-			if (e.Key == Keycode.V &&
-				((Platform.CurrentPlatform != PlatformType.OSX && e.Modifiers.HasModifier(Modifiers.Ctrl)) ||
-				 (Platform.CurrentPlatform == PlatformType.OSX && e.Modifiers.HasModifier(Modifiers.Meta))))
-			{
-				var clipboardText = Game.Renderer.GetClipboardText();
+				case Keycode.X:
+					ResetBlinkCycle();
+					if (((!isOSX && e.Modifiers.HasModifier(Modifiers.Ctrl)) ||
+						 (isOSX && e.Modifiers.HasModifier(Modifiers.Meta))) &&
+						 (!string.IsNullOrEmpty(Text)))
```

Comment:
Having the `||` and `&&` lined up is a bit misleading - the or actually applies to the first two groups, and the and to the result of that comparison and the third group.  You could make this a lot clearer by pulling out a temporary `cutHotkeyPressed` variable before the check.